### ISSUE TITLE
Azure function invocation failing with a-sync callback

### ISSUE THREAD

**Issue by CentricStraaten** [openfaas/faas#642]
I implemented OpenFaas on Azure AKS. 
I used https://github.com/openfaas/faas-netes as a source. All works fine. 
Only the a-sync call-back functionality doesn't work.

## Expected Behaviour

> I added a X-Callback-Url to the header with an URL to an Azure Function. 
> I expect the Function to call the callback URL when the routine ends.

## Current Behaviour

> The Azure Function works. 
> The OpenFaas function returns a when called a-sync 202 Status, but doens'nt call the Azure Function (callback URL)

## Possible Solution

> I have looked at the documentation on https://github.com/openfaas/faas/blob/master/guide/asynchronous.md. The problem is that i don't have the yml files mentioned in the instruction.

## Steps to Reproduce (for bugs)

> 1. http://cenaks01.westeurope.cloudapp.azure.com:8080/async-function/timefunction (expects an number in the body)
> 2. https://cencept-functions.azurewebsites.net/api/OpenFaasSMS?code=U1DZSiP0Ygcrh5anUgGdBQ705wz5LktE2AHtkaTqQCXGE80scnoHsw==&clientId=default&receiver=0612345678
> 3. curl --header X-Callback-Url: https://cencept-functions.azurewebsites.net/api/OpenFaasSMS?code=U1DZSiP0Ygcrh5anUgGdBQ705wz5LktE2AHtkaTqQCXGE80scnoHsw==&clientId=default&receiver=0653373439 --data 1 http://cenaks01.westeurope.cloudapp.azure.com:8080/async-function/timefunction

## Context

> I want to have a callback for async calls to check the result status.

## Your Environment

> Kubernetes (FaaS-netes) on Azure AKS

* Link to your project or a code example to reproduce issue:

* Please also follow the [troubleshooting guide](https://github.com/openfaas/faas/blob/master/guide/troubleshooting.md) and paste in any other diagnostic information you have:

**Comment by iyovcheva** [comment]
Derek edit title: Azure function invocation failing with a-sync callback

**Comment by CentricStraaten** [comment]
Hi, you should use a POST for the Azure Function. It expects a payload with data.

**Comment by CentricStraaten** [comment]
I don't have service logs from the Azure Functions, the function is never called. I have included a Postman JSON which you can import into Postman. I use these request to test both functions.

[OpenFaas.postman_collection.zip](https://github.com/openfaas/faas/files/1907393/OpenFaas.postman_collection.zip)

**Comment by CentricStraaten** [comment]
I solved the issue. The Azure Function did not accept a get. I changed the Function and the callback works!
Sorry, thank you for your support._

**Comment by iyovcheva** [comment]
Derek close

**Comment by alexellis** [comment]
When using curl any parameters or headers should be quoted with `""` i.e.

```
curl --header "X-Callback-Url: https://cencept-functions.azurewebsites.net/api/OpenFaasSMS?code=U1DZSiP0Ygcrh5anUgGdBQ705wz5LktE2AHtkaTqQCXGE80scnoHsw==&clientId=default&receiver=0653373439" --data "1" http://cenaks01.westeurope.cloudapp.azure.com:8080/async-function/timefunction -i
```

I can't see any code which is doing a GET here.

* The curl statement from Ivana is a post due to `--data 1`
* The x-callback-url is also going to invoke via a POST

The reason for seeing the error may have been because the `curl` was not quoted so bash interprets the text as other commands. 

Just wanted to make you aware that you could use an OpenFaaS function for the call-back too.

**Comment by CentricStraaten** [comment]
Why is the x-callback-url called multiple times (each 20 seconds as long as the function is executing)?

**Comment by CentricStraaten** [comment]
I am aware that we could use an OpenFaas function for call-back. For now it is convenient to use the Azure Function because it is allreay functional.

**Comment by alexellis** [comment]
What timeout did you configure for your asynchronous tasks? It sounds like you'll need to up the value. NATS Streaming is the system we use under the hood for asynchronous processing. It will retry work which doesn't complete in the deadline. I'd say this is what you're seeing. 

http://docs.openfaas.com/deployment/troubleshooting/#timeouts-asynchronous-invocations

**Comment by CentricStraaten** [comment]
I also updated the queue worker's setting

![image](https://user-images.githubusercontent.com/17175831/38980622-98ccb11c-43bd-11e8-8858-89f360c5ddd8.png)

**Comment by alexellis** [comment]
Try reproducing the issue in an isolated environment with a sleep function or similar.

**Comment by CentricStraaten** [comment]
What do you mean with an isolated environment? This is a POC environment which is only used by me.

**Comment by CentricStraaten** [comment]
I think I fixed it. I also had to change te env variables in the faas-netesd?

**Comment by alexellis** [comment]
@BurtonR can you offer any insights for @CentricStraaten ?

**Comment by BurtonR** [comment]
Hi @CentricStraaten! Were you able to resolve your issue? Could you share what you did, or when you realized what the problem was?

I'd recommend starting up a Play With Docker session (https://labs.play-with-docker.com/) and running the deployment one liner to get started in a fresh, clean environment. When you start a session, just follow the steps in the [Swarm Deployment Guide](https://docs.openfaas.com/deployment/docker-swarm/). This way you can be sure OpenFaaS and your functions work as expected before getting into more complex things.

If you need to test on Kubernetes you could try a local [Kubernetes deployment with MiniKube](https://medium.com/devopslinks/getting-started-with-openfaas-on-minikube-634502c7acdf). Just to make sure that there isn't some other process/app/networking issue causing your troubles

**Comment by CentricStraaten** [comment]
I added Application Insights to my function. In my function I added custom events to check if the could would run and finish after called a-sync. I used Postman to call my function on the async-function URL. I noticed that the function was initiated, received a timeout when the function exceeded the execution time of 8 seconds, and noticed that the code (sleep for 20 seconds) did not finish properly.

To increase the maximum run-time and avoid time-out exceptions i had to set the environment variables read-timeout and write_timeout in different containers. 

In the Azure AKS Kubernetes management portal I changed (or added) the variables in the yaml for following deployments (containers): 
- my function 
- gateway (see previous post)
- faas-netesd
 
In my case i raised the maximum time on the platform to 5 min, 300 seconds, and use different values per function. I tested long running functions successfully.

For the callback i added the X-Callback-Url variable in the header of the request. I noticed that the callback was called every 20 seconds.

In the queue-worker yaml I change ack_wait and ack_timeout from 20s to 300s also.

After this change the callback is only initiated once.
